transmit/direct_transmit_test.go

```diff
@@ -0,0 +1,502 @@
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			w.Write(packed)
+		}))
+		defer msgpackServer.Close()
+
+		dt, mockMetrics, _ := setupDirectTransmissionTest(t)
+		sendTestEvents(dt, msgpackServer.URL, 2, "test-api-key")
+
+		// Wait for events to be processed
+		assert.Eventually(t, func() bool {
+			success, _ := mockMetrics.Get(counterResponse20x)
+			errors, _ := mockMetrics.Get(counterResponseErrors)
+			return success == 1 && errors == 1
+		}, 100*time.Millisecond, 10*time.Millisecond)
+
+		// One success (202) and one error (400)
+		success, _ := mockMetrics.Get(counterResponse20x)
+		errors, _ := mockMetrics.Get(counterResponseErrors)
+		assert.Equal(t, float64(1), success)
+		assert.Equal(t, float64(1), errors)
+	})
+
+	t.Run("insufficient responses from server", func(t *testing.T) {
+		// Create a test server that returns fewer responses than events sent
+		insufficientServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			// Send only 1 response for 2 events
+			w.Write([]byte(`[{"status": 202}]`))
+		}))
+		defer insufficientServer.Close()
+
+		dt, mockMetrics, mockLogger := setupDirectTransmissionTest(t)
+		sendTestEvents(dt, insufficientServer.URL, 2, "test-api-key")
+
+		// Wait for events to be processed
+		assert.Eventually(t, func() bool {
+			success, _ := mockMetrics.Get(counterResponse20x)
+			errors, _ := mockMetrics.Get(counterResponseErrors)
+			return success == 1 && errors == 1
+		}, 100*time.Millisecond, 10*time.Millisecond)
+
+		// One success and one error (for missing response)
+		success, _ := mockMetrics.Get(counterResponse20x)
+		errors, _ := mockMetrics.Get(counterResponseErrors)
+		assert.Equal(t, float64(1), success)
+		assert.Equal(t, float64(1), errors)
+
+		// Verify error log message mentions insufficient responses
+		errorEvents := getErrorEvents(mockLogger)
+		require.Len(t, errorEvents, 1, "Expected exactly one error log for the missing response")
+
+		errorEvent := errorEvents[0]
+		assert.Equal(t, "error when sending event", errorEvent.Fields["error"])
+		assert.Equal(t, http.StatusInternalServerError, errorEvent.Fields["status_code"])
+		assert.Contains(t, errorEvent.Fields, "roundtrip_usec")
+	})
+}
+
+func TestDirectTransmission(t *testing.T) {
+	testServer := newTestDirectAPIServer(t)
+
+	mockMetrics := &metrics.MockMetrics{}
+	mockMetrics.Start()
+
+	// Use max batch size of 3 for testing
+	dt := NewDirectTransmission(mockMetrics, "test", 3, 50*time.Millisecond)
+	dt.Logger = &logger.NullLogger{}
+	dt.Version = "test-version"
+	dt.Transport = http.DefaultTransport.(*http.Transport)
+
+	err := dt.Start()
+	require.NoError(t, err)
+	defer dt.Stop()
+
+	now := time.Now().UTC()
+
+	// Create events for multiple datasets and scenarios
+	var allEvents []*types.Event
+
+	// Dataset A: 5 events (should create 2 batches: 3 + 2)
```

Comment:
Is there a way to verify the batching logic works as expected as well?

Reply:
I added an assertion to the test server that the batch size doesn't exceed the maximum.